File: templates/problems/18_bit_manipulation/06_sum_two_integers/typescript/solution.ts
/**
 * Add two integers using bit manipulation. O(1) time (bounded iterations), O(1) space.
 * JavaScript bitwise operators work on 32-bit signed integers, so the carry
 * naturally shifts out after 32 bits.
 */
export function solveGetSum(a: number, b: number): number {
    let aa = a, bb = b;
    while (bb !== 0) {
        const carry = aa & bb;
        aa = aa ^ bb;
        bb = carry << 1;
    }
    return aa;
}

File: templates/problems/18_bit_manipulation/06_sum_two_integers/go/solution.go
package sum_two_integers

// solveGetSum adds two integers using bit manipulation. O(1) time (bounded iterations), O(1) space.
func solveGetSum(a int, b int) int {
	for b != 0 {
		carry := a & b
		a = a ^ b
		b = carry << 1
	}
	return a
}
